frameworks/Rust/actix/src/db_pg_direct.rs

```diff
@@ -0,0 +1,154 @@
+                                            update
+                                        })
+                                    })
+                            })
+                    })
+            })
+    }
+}
+
+impl PgConnection {
+
+    pub fn get_world(&self) -> impl Future<Item=World, Error=io::Error> {
+        let random_id = thread_rng().gen_range::<i32>(1, 10_000);
+
+        self.cl
+            .query(&self.world, &[&random_id])
+            .into_future()
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e.0))
+            .and_then(|(row, _)| {
+                let row = row.unwrap();
+                Ok(World {
+                    id: row.get(0),
+                    randomnumber: row.get(1),
+                })
+            })
+    }
+
+    pub fn get_worlds(&self, num: usize) -> impl Future<Item=Vec<World>, Error=io::Error> {
+        let mut worlds = Vec::with_capacity(num);
+        for _ in 0..num {
+            let w_id: i32 = thread_rng().gen_range(1, 10_000);
+            worlds.push(
+                self.cl
+                    .query(&self.world, &[&w_id])
+                    .into_future()
+                    .map_err(|e| io::Error::new(io::ErrorKind::Other, e.0))
+                    .and_then(|(row, _)| {
+                        let row = row.unwrap();
+                        Ok(World {
+                            id: row.get(0),
+                            randomnumber: row.get(1),
+                        })
+                    }),
+            );
+        }
+
+        stream::futures_unordered(worlds).collect()
+    }
+
+    pub fn update(&self, num: usize) -> impl Future<Item=Vec<World>, Error=io::Error> {
+        let mut worlds = Vec::with_capacity(num);
+        for _ in 0..num {
+            let id: i32 = thread_rng().gen_range(1, 10_000);
+            let w_id: i32 =thread_rng().gen_range(1, 10_000);
+            worlds.push(
+                self.cl
+                    .query(&self.update, &[&w_id])
+                    .into_future()
+                    .map_err(|e| io::Error::new(io::ErrorKind::Other, e.0))
+                    .and_then(move |(row, _)| {
+                        let row = row.unwrap();
+                        Ok(World {
+                            id: row.get(0),
+                            randomnumber: id,
+                        })
+                    }),
+            );
+        }
+
+        let cl = self.cl.clone();
+        stream::futures_unordered(worlds)
+            .collect()
+            .and_then(move |mut worlds| {
+                let mut update = String::with_capacity(120 + 6 * num as usize);
+                update
+                    .push_str("UPDATE world SET randomnumber = temp.randomnumber FROM (VALUES ");
+
+                for w in &worlds {
+                    update.push_str(&format!("({}, {}),", w.id, w.randomnumber));
+                }
+                worlds.sort_by_key(|w| w.id);
```

Comment:
I think you can remove this sort operation.  The `ORDER BY 1` in your update statement should take care of the batch update deadlock issue, since that sorts the `VALUES` by id.  (My theory about this code, which may be wrong, is that `worlds.sort_by_key` was originally added to this code to address deadlocks, but now it has drifted into the wrong place in the code.  It would have to be before the `for w in &worlds` loop to have its intended effect, but it's redundant because of the `ORDER BY 1` anyway.)

Reply:
i am not sure about that, i remember there was some problems with query without this sort. i will test again.